Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -710,25 +710,129 @@ WarpXParticleContainer::DepositCharge (WarpXParIter& pti, RealVector& wp,
+    {
+        const Geometry& geom = Geom(lev);
+        const auto dxi = geom.InvCellSizeArray();
+        const auto plo = geom.ProbLoArray();
+        const auto domain = geom.Domain();
+
+        auto& ptile = ParticlesAt(lev, pti);
+        auto& aos   = ptile.GetArrayOfStructs();
+        auto pstruct_ptr = aos().dataPtr();
+
+        Box box = pti.validbox();
+        box.grow(ng_rho);
+        amrex::IntVect bin_size = WarpX::sort_bin_size;
+
+        const auto offsets_ptr = bins.offsetsPtr();
+        auto tbox_ptr = tboxes.dataPtr();
+        amrex::ParallelFor(bins.numBins(),
+                           [=] AMREX_GPU_DEVICE (int ibin) {
+                               const int bin_start = offsets_ptr[ibin];
+                               const int bin_stop = offsets_ptr[ibin+1];
+                               if (bin_start < bin_stop) {
+                                   const auto& p = pstruct_ptr[bin_start];
+                                   Box tbx;
+                                   auto iv = getParticleCell(p, plo, dxi, domain);
+                                   AMREX_ALWAYS_ASSERT(box.contains(iv)); // remove later
+                                   auto tid = getTileIndex(iv, box, true, bin_size, tbx);
+                                   AMREX_ALWAYS_ASSERT(tid == ibin); // remove later
+                                   AMREX_ALWAYS_ASSERT(tbx.contains(iv)); // remove later
+                                   tbox_ptr[ibin] = tbx;
+                               }
+                           });
+    }
+
+    // compute max tile box size in each direction
+    amrex::IntVect max_tbox_size;
+    {
+        ReduceOps<AMREX_D_DECL(ReduceOpMax, ReduceOpMax, ReduceOpMax)> reduce_op;
+        ReduceData<AMREX_D_DECL(int, int, int)> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+
+        const auto boxes_ptr = tboxes.dataPtr();
+        reduce_op.eval(tboxes.size(), reduce_data,
+        [=] AMREX_GPU_DEVICE (int i) -> ReduceTuple
+            {
+                const Box& box = boxes_ptr[i];
+                if (box.ok()) {
+                  IntVect si = box.length();
+                  return {AMREX_D_DECL(si[0], si[1], si[2])};
+                } else {
+                  return {AMREX_D_DECL(0, 0, 0)};
+                }
+            });
+
+        ReduceTuple hv = reduce_data.value();
+
+        max_tbox_size = IntVect(AMREX_D_DECL(amrex::get<0>(hv),
+                                             amrex::get<1>(hv),
+                                             amrex::get<2>(hv)));
+    }
+
     WARPX_PROFILE_VAR_START(blp_ppc_chd);
     amrex::LayoutData<amrex::Real>* costs = WarpX::getCosts(lev);
     amrex::Real* cost = costs ? &((*costs)[pti.index()]) : nullptr;
 
-    if        (WarpX::nox == 1){
+    const auto GetPosition = GetParticlePosition(pti, offset);
+    const Geometry& geom = Geom(lev);
+    Box box = pti.validbox();
+    box.grow(1);
```

Comment:
Does this also need to be `box.grow(ng_rho)`? @atmyers

Reply:
I believe so, because we want the result of the `getTileIndex` function call to always be same inside and outside of the `doChargeDeposition` function call. That is the only thing this `box` is used for.